### PR TITLE
Post details - #2 Fetching data

### PR DESCRIPTION
**Pull request by matej494**

### CODE REVIEW

File: Marsiple/Common/Network/DataFetcher.swift [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+                            .appendingPathComponent(MartianApi.URLs.posts)
+                            .appendingPathComponent("\(id)")
+                            .appendingPathComponent(MartianApi.URLs.comments)
+            else { return DispatchQueue.main.async { failure(DataFetcherError.urlCreationFailure) } }
+        getData(url: url, method: HTTPRequestMethod.get, success: success, failure: failure)
+    }
+}
+
+private extension DataFetcher {
+    static func getData<DataType: Codable>(url: URL,
```
Comment (dkulundzic) [inline Marsiple/Common/Network/DataFetcher.swift:30]: Woot, generics! 🎉

File: Marsiple/Common/Network/DataFetcher.swift [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+                                           method: String,
+                                           success: @escaping ([DataType]) -> Void,
+                                           failure: @escaping (LocalizedError) -> Void) {
+        var request = URLRequest(url: url)
+        request.httpMethod = method
+        request.addValue(MartianApi.Headers.contentTypeValue, forHTTPHeaderField: HTTPRequestHeader.contentType)
```
Comment (dkulundzic) [inline Marsiple/Common/Network/DataFetcher.swift]: Just noticed this one - I think we could name this one better than it currently is. Basically, what we're doing here, is we're extracting the `default` value for our `Content-Type` header. 

Thus, we could use `MartianApi.Headers.contentTypeValue` as `MartianApi.Headers.defaultContentType` or similiar. What do you think?

File: Marsiple/Common/Network/DataFetcher.swift [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        request.addValue(MartianApi.Headers.xAuthValue, forHTTPHeaderField: HTTPRequestHeader.xAuth)
+        
+        let session = URLSession.shared
+        let task = session.dataTask(with: request) { data, response, error in
```
Comment (dkulundzic) [inline Marsiple/Common/Network/DataFetcher.swift]: Is it me, or we're not using the `error` argument anywhere? We should definitely do a `guard` on the error and do an early return if we've encountered and `error`.

File: Marsiple/Common/Network/DataFetcher.swift [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+//  DataFetcher.swift
+//  Marsiple
+//
+//  Created by Matej Korman on 23/07/2018.
+//  Copyright © 2018 Matej Korman. All rights reserved.
+//
+
+import Foundation
+
+class DataFetcher {
```
Comment (dkulundzic) [inline Marsiple/Common/Network/DataFetcher.swift:11]: I like what you did here 👍 

Do you see a possibility for making the `DataFetcher` fully generic?

**Review by dkulundzic — CHANGES_REQUESTED** [review]
I massively approve of the use of generics here, woohoo ! 

Minor nitpicks.

File: Marsiple/Common/Network/DataFetcher.swift (hunk 2 above)
Comment (matej494) [reply Marsiple/Common/Network/DataFetcher.swift]: Ok. Looks more descriptive.

File: Marsiple/Common/Network/DataFetcher.swift (hunk 4 above)
Comment (matej494) [reply Marsiple/Common/Network/DataFetcher.swift:11]: You mean like this `class DataFetcher<DataType: Codable>`?

Comment (dkulundzic) [reply Marsiple/Common/Network/DataFetcher.swift:11]: I do, yeah.

File: Marsiple/Common/Network/DataFetcher.swift [hunk 5]
```diff
@@ -33,11 +33,14 @@ private extension DataFetcher {
...
         request.httpMethod = method
-        request.addValue(MartianApi.Headers.contentTypeValue, forHTTPHeaderField: HTTPRequestHeader.contentType)
+        request.addValue(MartianApi.Headers.defaultContentType, forHTTPHeaderField: HTTPRequestHeader.contentType)
         request.addValue(MartianApi.Headers.xAuthValue, forHTTPHeaderField: HTTPRequestHeader.xAuth)
         
         let session = URLSession.shared
         let task = session.dataTask(with: request) { data, response, error in
+            if let error = error {
+                return DispatchQueue.main.async { failure(DataFetcherError.generic(error)) }
+            }
```
Comment (dkulundzic) [inline Marsiple/Common/Network/DataFetcher.swift]: Lets add a newline here, will be a bit more readable.

File: Marsiple/Common/Network/DataFetcher.swift [hunk 6]
```diff
@@ -33,11 +33,14 @@ private extension DataFetcher {
                                            failure: @escaping (LocalizedError) -> Void) {
         var request = URLRequest(url: url)
         request.httpMethod = method
-        request.addValue(MartianApi.Headers.contentTypeValue, forHTTPHeaderField: HTTPRequestHeader.contentType)
+        request.addValue(MartianApi.Headers.defaultContentType, forHTTPHeaderField: HTTPRequestHeader.contentType)
         request.addValue(MartianApi.Headers.xAuthValue, forHTTPHeaderField: HTTPRequestHeader.xAuth)
         
         let session = URLSession.shared
         let task = session.dataTask(with: request) { data, response, error in
+            if let error = error {
```
Comment (dkulundzic) [inline Marsiple/Common/Network/DataFetcher.swift]: Great 👍

File: Marsiple/Common/Network/DataFetcher.swift (hunk 4 above)
Comment (matej494) [reply Marsiple/Common/Network/DataFetcher.swift:11]: After thinking about it for a while yes, but I'm not quite sure how to do this. I think i need just a bit more time to figure it out.

Comment (dkulundzic) [reply Marsiple/Common/Network/DataFetcher.swift:11]: This is not a trivial solution, but is a good example of what it takes to trully make generics work. 

We can leave this as an improvement in the future, you don't have to be blocked by this.

**Review by dkulundzic — APPROVED** [review]

**Comment by matej494** [comment]
Successfully merged branch `post-detail/fetching-data` into branch `post-details/main` after resolving conflicts.

**Pull request merged**